File: tools/ui-test-agent/src/browsertest.ts
/**
 * Real-browser acceptance test runner for the ASDLC pipeline.
 *
 * Executes an LLM-authored test plan (JSON) against a running app with a real
 * Chromium browser and reports the actual outcome of each step. Deterministic:
 * the plan says what to do and expect; this only drives the browser and judges
 * literal outcomes (visible text, URL, element count, axe) — no per-step LLM.
 *
 * HEADED=1 opens a visible slow-motion browser (the standing "watch the test"
 * rule). A screenshot per case is captured for the report.
 *
 * Usage:
 *   npm run browsertest -- --base-url http://localhost:8088 \
 *     --plan plan.json --out results.json --shots ./shots
 *
 * Plan shape:
 *   { cases: [ { id, title, steps: [ Step, ... ] } ] }
 * Step (one key each):
 *   { goto: "/path" }
 *   { click: "Button text" }                 // by accessible name / text
 *   { fill: { selector|label: "...", value: "..." } }
 *   { expectText: "visible text" }            // present anywhere on the page
 *   { expectUrl: "/path" }                    // current path contains this
 *   { expectCount: { selector: "css", min?: n, max?: n } }
 *   { expectNoAxe: true }                     // no serious/critical violations
 */

import { chromium, type Page } from "playwright";
import AxeBuilder from "@axe-core/playwright";
import * as fs from "fs";
import * as path from "path";

interface Step { [k: string]: any }
interface Case { id: string; title: string; steps: Step[] }
interface StepResult { step: string; ok: boolean; detail: string }
interface CaseResult {
  id: string; title: string; ok: boolean; screenshot: string; steps: StepResult[];
}

function arg(name: string, fallback = ""): string {
  const i = process.argv.indexOf(`--${name}`);
  return i >= 0 && process.argv[i + 1] ? process.argv[i + 1] : fallback;
}

const baseUrl = arg("base-url", "http://localhost:3000").replace(/\/$/, "");
const planFile = arg("plan");
const outFile = arg("out", "browsertest-results.json");
const shotsDir = arg("shots", "shots");
const headed = !!process.env.HEADED;

const label = (step: Step): string => JSON.stringify(step).slice(0, 120);

async function runStep(page: Page, step: Step): Promise<StepResult> {
  const key = Object.keys(step)[0];
  try {
    if ("goto" in step) {
      await page.goto(baseUrl + step.goto, { waitUntil: "networkidle" });
    } else if ("click" in step) {
      await page.getByText(step.click, { exact: false }).first().click({ timeout: 8000 });
    } else if ("fill" in step) {
      const f = step.fill;
      const loc = f.label
        ? page.getByLabel(f.label, { exact: false })
        : page.locator(f.selector);
      await loc.first().fill(String(f.value), { timeout: 8000 });
    } else if ("expectText" in step) {
      await page.getByText(step.expectText, { exact: false }).first()
        .waitFor({ state: "visible", timeout: 8000 });
    } else if ("expectUrl" in step) {
      await page.waitForTimeout(300);
      const u = new URL(page.url());
      if (!(u.pathname + u.hash).includes(step.expectUrl)) {
        return { step: label(step), ok: false, detail: `url is ${u.pathname + u.hash}` };
      }
    } else if ("expectCount" in step) {
      const c = step.expectCount;
      const n = await page.locator(c.selector).count();
      if (c.min != null && n < c.min) return { step: label(step), ok: false, detail: `count ${n} < min ${c.min}` };
      if (c.max != null && n > c.max) return { step: label(step), ok: false, detail: `count ${n} > max ${c.max}` };
    } else if ("expectNoAxe" in step) {
      const res = await new AxeBuilder({ page }).analyze();
      const serious = res.violations.filter((v) => ["serious", "critical"].includes(v.impact ?? ""));
      if (serious.length) return { step: label(step), ok: false, detail: `axe: ${serious.map((v) => v.id).join(", ")}` };
    } else {
      return { step: label(step), ok: false, detail: `unknown step key "${key}"` };
    }
    return { step: label(step), ok: true, detail: "" };
  } catch (e: any) {
    return { step: label(step), ok: false, detail: (e?.message ?? String(e)).slice(0, 160) };
  }
}

async function main(): Promise<void> {
  const plan = JSON.parse(fs.readFileSync(planFile, "utf-8"));
  const cases: Case[] = plan.cases ?? [];
  fs.mkdirSync(shotsDir, { recursive: true });

  const browser = await chromium.launch(headed ? { headless: false, slowMo: 300 } : {});
  const ctx = await browser.newContext({ viewport: { width: 1280, height: 900 } });
  const results: CaseResult[] = [];

  for (const c of cases) {
    const page = await ctx.newPage();
    const steps: StepResult[] = [];
    for (const step of c.steps) {
      const r = await runStep(page, step);
      steps.push(r);
      if (!r.ok) break; // stop the case at first failing step
    }
    const shot = path.join(shotsDir, `${c.id}.png`);
    await page.screenshot({ path: shot }).catch(() => {});
    results.push({
      id: c.id, title: c.title, ok: steps.every((s) => s.ok),
      screenshot: shot, steps,
    });
    await page.close();
  }

  await browser.close();
  const passed = results.filter((r) => r.ok).length;
  fs.writeFileSync(outFile,
    JSON.stringify({ base_url: baseUrl, total: results.length, passed, cases: results }, null, 2) + "\n");
  console.log(`browsertest: ${passed}/${results.length} cases passed`);
  process.exit(0);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: pipeline/agents/browsertest.py
"""Browser test agent: author real-browser acceptance tests from the backlog,
run them against the live app, report, and file failures as PROBLEM tickets.

On-demand stage (run-pipeline.py run --stage browsertest). The chain so far:
testgen writes Gherkin specs, qa runs them for the gate. This agent goes
further — the LLM reads the acceptance criteria and AUTHORS a concrete browser
test plan (navigate/click/fill + expect visible text / URL / count / axe),
which a real Chromium browser executes step-by-step (tools/ui-test-agent
browsertest mode, HEADED-capable). Actual outcomes are judged deterministically.

Failing cases are LLM-triaged into 01_requirements/discovered/tickets/PROBLEM-*
so the self-improvement loop can pick them up. The stage reports pass/fail but
does not hard-fail the pipeline (finding gaps is useful output).

Target: BROWSERTEST_URL or AUDIT_URL if set (e.g. http://localhost:8088),
otherwise dev servers are booted locally like the QA stage.
"""

from __future__ import annotations

import json
import logging
import os
import re
import shutil
import subprocess

from . import register
from .base import Agent, AgentResult
from .qa import DevServer

log = logging.getLogger("pipeline")

PLAN_SCHEMA = {
    "type": "object",
    "properties": {
        "cases": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "id": {"type": "string"},          # STORY-001-c1
                    "title": {"type": "string"},
                    "story": {"type": "string"},
                    "steps": {
                        "type": "array",
                        "items": {"type": "object"},     # one action/expect per step
                    },
                },
                "required": ["id", "title", "steps"],
                "additionalProperties": False,
            },
        },
    },
    "required": ["cases"],
    "additionalProperties": False,
}

PLAN_PROMPT = """You are the Browser Test agent in an automated software delivery
pipeline. From the product backlog (INVEST stories + Gherkin acceptance
criteria) and the app's routes, author a concrete browser test PLAN a real
Chromium browser will execute against the running app.

Output JSON: { "cases": [ { "id", "title", "story", "steps": [...] } ] }.
- One case per acceptance scenario. id like "STORY-001-c1".
- Each step is an object with exactly ONE key. Allowed steps:
  { "goto": "/path" }                         navigate (path only, no host)
  { "click": "Visible button or link text" }  click by visible text
  { "fill": { "label": "Field label", "value": "..." } }  fill an input by label
  { "expectText": "text that must be visible" }
  { "expectUrl": "/path" }                     current path/hash contains this
  { "expectCount": { "selector": "css", "min": 1 } }  element count bound
  { "expectNoAxe": true }                      no serious/critical a11y violations
- Start each case with a goto. Derive expected copy VERBATIM from the Gherkin
  (visible text, URLs). Prefer expectText/expectUrl over brittle CSS.
- Only test what the stories specify. Keep each case to 3-8 steps.
- Include one { "expectNoAxe": true } case per user-facing page."""

TRIAGE_SCHEMA = {
    "type": "object",
    "properties": {
        "tickets": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "title": {"type": "string"},
                    "severity": {"type": "string", "enum": ["P0", "P1", "P2"]},
                    "area": {"type": "string"},
                    "problem": {"type": "string"},
                    "gherkin": {"type": "string"},
                },
                "required": ["title", "severity", "area", "problem", "gherkin"],
                "additionalProperties": False,
            },
        },
        "summary": {"type": "string"},
    },
    "required": ["tickets", "summary"],
    "additionalProperties": False,
}

TRIAGE_PROMPT = """You are the Browser Test agent triaging FAILED browser test
cases into actionable problem tickets. Each failure has the case title, the
step that failed, and the browser's actual detail. Merge duplicates (same root
cause across cases = ONE ticket). One ticket per coherent defect with a Gherkin
reproduction (Given/When/Then). severity: P0 = a primary user flow is broken;
P1 = a story's acceptance criterion fails or a WCAG violation; P2 = minor.
`summary`: 2-3 sentences on how well the app meets its acceptance criteria."""


@register("browsertest")
class BrowserTestAgent(Agent):
    extra_write_roots = ("01_requirements",)

    def run(self) -> AgentResult:
        root = self.ctx.root
        backlog_path = root / "02_specs" / "PRODUCT_BACKLOG.md"
        if not backlog_path.exists():
            return AgentResult(ok=False, summary="02_specs/PRODUCT_BACKLOG.md missing")
        backlog = backlog_path.read_text(encoding="utf-8")

        # 1. LLM authors the browser test plan from the acceptance criteria
        routes = self._app_routes()
        plan_res = self.ctx.llm.complete(
            system=self.system_blocks(PLAN_PROMPT),
            user=f"App routes: {', '.join(routes)}\n\nProduct backlog:\n\n{backlog}",
            schema=PLAN_SCHEMA,
            max_tokens=12000,
        )
        plan = plan_res.parsed
        self.write_file("05_test_reports/browser/plan.json",
                        json.dumps(plan, indent=2) + "\n")
        if not plan["cases"]:
            return AgentResult(ok=True, usage=plan_res.usage,
                               summary="no test cases authored")

        # 2. a real browser executes the plan against the running app
        servers: list[DevServer] = []
        target = os.environ.get("BROWSERTEST_URL") or os.environ.get("AUDIT_URL")
        try:
            if not target:
                backend = root / "04_source" / "backend"
                if backend.exists():
                    servers.append(DevServer(backend, "http://localhost:3001/api/health", "backend"))
                servers.append(DevServer(root / "04_source" / "frontend",
                                         "http://localhost:3000", "frontend"))
                target = "http://localhost:3000"
            results = self._run_browser(target)
        except RuntimeError as exc:
            return AgentResult(ok=False, summary=f"browser test target unavailable: {exc}")
        finally:
            for server in servers:
                server.stop()

        if results is None:
            return AgentResult(ok=False, summary="browser runner failed (see logs)")

        passed = results["passed"]
        total = results["total"]
        failures = [c for c in results["cases"] if not c["ok"]]

        # 3. triage failures into PROBLEM tickets (like audit mode)
        written: list[str] = []
        summary_text = f"{passed}/{total} browser acceptance cases passed"
        usage = plan_res.usage
        if failures:
            fail_digest = [{
                "title": c["title"],
                "failed_step": next((s["step"] for s in c["steps"] if not s["ok"]), ""),
                "actual": next((s["detail"] for s in c["steps"] if not s["ok"]), ""),
            } for c in failures]
            triage = self.ctx.llm.complete(
                system=self.system_blocks(TRIAGE_PROMPT),
                user=f"Target: {target}\n\nFailed cases (JSON):\n\n"
                     f"```json\n{json.dumps(fail_digest, indent=2)[:16000]}\n```",
                schema=TRIAGE_SCHEMA,
                max_tokens=8000,
            )
            usage = (usage or triage.usage)
            written = self._write_tickets(triage.parsed["tickets"])
            summary_text += f"; {triage.parsed['summary']}"

        self._write_report(target, results, written)
        return AgentResult(
            ok=True, usage=usage,
            summary=f"{passed}/{total} cases passed"
                    + (f", {len(failures)} failing -> {len(written)} ticket(s)" if failures else ""),
        )

    # -- browser runner -----------------------------------------------------

    def _run_browser(self, target: str) -> dict | None:
        root = self.ctx.root
        ui_agent = root / "tools" / "ui-test-agent"
        if not (ui_agent / "node_modules").exists():
            log.warning("browsertest: ui-test-agent not installed — skipping")
            return None
        plan = root / "05_test_reports" / "browser" / "plan.json"
        out = root / "05_test_reports" / "browser" / "results.json"
        shots = root / "05_test_reports" / "browser" / "shots"
        npm = shutil.which("npm") or "npm"
        proc = subprocess.run(
            [npm, "run", "browsertest", "--", "--base-url", target,
             "--plan", str(plan), "--out", str(out), "--shots", str(shots)],
            cwd=ui_agent, capture_output=True, text=True,
            encoding="utf-8", errors="replace", timeout=900,
        )
        if not out.exists():
            log.warning("browsertest: runner produced no results: %s",
                        (proc.stderr or proc.stdout)[-400:])
            return None
        return json.loads(out.read_text(encoding="utf-8"))

    def _app_routes(self) -> list[str]:
        app = self.ctx.root / "04_source" / "frontend" / "src" / "App.tsx"
        if not app.exists():
            return ["/"]
        routes = re.findall(r'path="([^"*]+)"', app.read_text(encoding="utf-8"))
        return sorted(set(routes) | {"/"})

    # -- output (ticket writer mirrors the audit agent) --------------------

    def _write_tickets(self, tickets: list[dict]) -> list[str]:
        tickets_dir = self.ctx.root / "01_requirements" / "discovered" / "tickets"
        deferred_dir = self.ctx.root / "01_requirements" / "discovered" / "deferred"
        existing = [int(m.group(1)) for p in tickets_dir.glob("PROBLEM-*.md")
                    if (m := re.match(r"PROBLEM-(\d+)", p.name))] if tickets_dir.exists() else []
        seen: set[str] = set()
        for folder in (tickets_dir, deferred_dir):
            if folder.exists():
                for p in folder.glob("PROBLEM-*.md"):
                    seen.add(re.sub(r"[^a-z0-9]", "",
                                    p.read_text(encoding="utf-8").splitlines()[0].lower()))
        next_id = max(existing, default=0) + 1
        written = []
        for t in tickets:
            key = re.sub(r"[^a-z0-9]", "", f"# {t['severity']} — {t['title']}".lower())
            if key in seen:
                continue
            name = f"PROBLEM-{next_id:03d}"
            self.write_file(
                f"01_requirements/discovered/tickets/{name}.md",
                f"# {t['severity']} — {t['title']}\n\n**Area:** {t['area']}\n\n"
                f"{t['problem']}\n\n## Acceptance criteria\n\n"
                f"```gherkin\n{t['gherkin'].strip()}\n```\n",
            )
            written.append(name)
            next_id += 1
        return written

    def _write_report(self, target: str, results: dict, tickets: list[str]) -> None:
        lines = [f"# Browser Acceptance Test — {target}", "",
                 f"**{results['passed']}/{results['total']} cases passed.**", ""]
        for c in results["cases"]:
            mark = "✅" if c["ok"] else "❌"
            lines.append(f"### {mark} {c['id']} — {c['title']}")
            if not c["ok"]:
                bad = next((s for s in c["steps"] if not s["ok"]), None)
                if bad:
                    lines.append(f"- failed at `{bad['step']}` — {bad['detail']}")
            lines.append(f"- screenshot: `{c['screenshot']}`")
            lines.append("")
        if tickets:
            lines += ["## Filed problem tickets", "",
                      *[f"- `01_requirements/discovered/tickets/{t}.md`" for t in tickets],
                      "", "Re-run the product stage to fold them into the backlog."]
        self.write_file("05_test_reports/browser/BROWSER_TEST.md", "\n".join(lines) + "\n")
